### PR TITLE
[WIP] Move float8 cutlass sparse layout to Float8SemiSparseTensor

### PR DESCRIPTION
**Pull request by bbeckca**
Moving float8 cutlass sparse layout into its own class:
https://github.com/pytorch/ao/blob/main/torchao/dtypes/floatx/cutlass_semi_sparse_layout.py

cc @jerryzh168 @jcaip @danielvegamyhre

### CODE REVIEW

File: torchao/quantization/quantize_/common/packing_format.py [hunk 1]
```diff
@@ -32,3 +32,4 @@ class PackingFormat(str, Enum):
     needed for the rest of the system to understand the specific format that's adopted.
     """
     OPAQUE = "opaque"
+    # todo: add semi-sparse
```
Comment (bbeckca) [inline torchao/quantization/quantize_/common/packing_format.py]: @jerryzh168 It seems we may want to add a packing format for sparse. Wondering if there's a preference between adding it here or in a separate file ([similar to int4](https://github.com/pytorch/ao/blob/30082cb3d0e6f86b1bae5e75ef4284e3a4b5ece6/torchao/quantization/quantize_/workflows/int4/int4_packing_format.py#L4)) for float8?

Comment (jcaip) [reply torchao/quantization/quantize_/common/packing_format.py]: Do we need packing format if we have a separate config? It looks like packing format is mostly to support different Int4WeightOnlyConfig kernel options (tinygemm, sparse marlin, etc).

File: test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+from torchao.quantization import (
+    Float8WeightOnlyConfig,
+    quantize_,
+)
+from torchao.quantization.utils import compute_error
+from torchao.sparsity.sparse_api import apply_fake_sparsity
+from torchao.testing.utils import skip_if_rocm
+from torchao.utils import torch_version_at_least
+
+BF16_ACT_CONFIG = Float8WeightOnlyConfig(
```
Comment (jcaip) [inline test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py]: I don't think this config makes sense, it's not something we support. From what I understand this is a bf16 a + fp8 sparse weight? We only have kernel support for fp8xfp8 +2:4 sparse matmul, no support for mixed  input dtypes currently.

File: torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            meta,
+            scale,
+        )
+
+
+implements = Float8SemiSparseTensor.implements
+implements_torch_function = Float8SemiSparseTensor.implements_torch_function
+
+
+@implements(aten.linear.default)
```
Comment (jcaip) [inline torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py]: We'll also need to make sure `mm` and `addmm` are supported ops as well. The arg order is different from linear but it should be the same logic.

Comment (bbeckca) [reply torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py]: Sounds good, I'm onboard with that. Mind if I add those ops in a follow-up diff after this lands?

File: torchao/quantization/quantize_/common/packing_format.py (hunk 1 above)
Comment (bbeckca) [reply torchao/quantization/quantize_/common/packing_format.py]: Good point, I noticed that we seem to replace the dense weight with quantized semi-sparse [in the transform](https://github.com/pytorch/ao/blob/ed4cd34323ce58c4e82d1561a72abee3356eb1ad/torchao/quantization/quant_api.py#L1881) Would it make more sense to integrate Float8SemiSparseTensor here rather than gating with packing-format as I proposed previously? cc @jerryzh168

File: test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py (hunk 2 above)
Comment (bbeckca) [reply test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py]: You're right, it seems I should be mirroring test_fp8_cutlass_sparse ([from test_sparse_api.py](https://github.com/pytorch/ao/blob/ed4cd34323ce58c4e82d1561a72abee3356eb1ad/test/sparsity/test_sparse_api.py#L132)) instead
with the difference being using the new flag/config which exposes the tensor subclass being added?

File: torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py (hunk 3 above)
Comment (jcaip) [reply torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py]: Yup that's fine with me :)

File: test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py (hunk 2 above)
Comment (jcaip) [reply test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py]: I think `Float8DynamicActivationFloat8SemiSparseWeightConfig` should eventually resolve to your subclass.
But I would like to hold off on that until 
1) We're out of the QRT period (cc @RandySheriff just FYI on refactor plans)
2) We have the same functionality (addmm/mm support)

File: torchao/quantization/quantize_/common/packing_format.py (hunk 1 above)
Comment (jcaip) [reply torchao/quantization/quantize_/common/packing_format.py]: Yup I think that the transform should call your subclass.

File: test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py (hunk 2 above)
Comment (bbeckca) [reply test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py]: Sounds good. In that case, should we sequence the changes as follows?
1. Land `Float8SemiSparseTensor` with linear support
2. Add mm / addmm ops for feature parity
3. Integrate into `Float8DynamicActivationFloat8SemiSparseWeightConfig` after QRT

Comment (bbeckca) [reply test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py]: @jcaip @jerryzh168 Mind confirming that you're onboard with this direction?

Comment (jcaip) [reply test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py]: Yeah, that sounds good to me.

File: test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+    @skip_if_rocm("ROCm enablement in progress")
+    @parametrize(
+        "sizes",
+        [
+            ((128,), 256, 128),
+            ((32, 128), 512, 128),
+            ((2, 32, 128), 256, 128),
+        ],
+    )
+    def test_sparse_vs_dense_fp8(self, sizes):
```
Comment (bbeckca) [inline test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py:42]: @jcaip Updated testing to follow style of [test_sparse_apu.py](https://fburl.com/18n157bf. For now I'm omitting any config related changes until QRT, however this diff does include all ops (linear, addmm, mm) so that integration can be done as follow up.

Could I get feedback on the construction of the implementations and test before adding similar for addmm, mm and adding polish?

File: test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        input_fp8 = Float8Tensor.from_hp(
+            input, float8_dtype=torch.float8_e4m3fn, mm_config=mm_config
+        )
+
+        weight_fp8 = Float8Tensor.from_hp(
+            linear.weight.data, float8_dtype=torch.float8_e4m3fn, mm_config=mm_config
+        )
+        dense_output = torch.nn.functional.linear(input_fp8, weight_fp8, linear.bias)
+
+        weight_sparse_fp8 = Float8SemiSparseTensor.from_hp(linear.weight.data, [1, K])
```
Comment (jcaip) [inline test/quantization/quantize_/workflows/float8/test_float8_semi_sparse.py:62]: nit: .detach() instead of .data?

File: torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+    _quantize_affine_float8,
+)
+from torchao.utils import TorchAOBaseTensor
+
+__all__ = ["Float8SemiSparseTensor"]
+aten = torch.ops.aten
+
+
+class Float8SemiSparseTensor(TorchAOBaseTensor):
+    tensor_data_names = ["sparse", "meta", "scale"]
```
Comment (jcaip) [inline torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py:22]: nit: we should use [ `compressed_values`, `metadata` ]  instead of `sparse` and `meta` here.

File: torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+            meta,
+            scale,
+        )
+
+
+implements = Float8SemiSparseTensor.implements
+implements_torch_function = Float8SemiSparseTensor.implements_torch_function
+
+
+@implements(aten.t.default)
```
Comment (jcaip) [inline torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py:86]: Why do you have to implement transpose? Transpose on a sparse matrix is kind of tricky, we should probably throw an `ValueError` if it's called.

File: torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+    return return_and_correct_aliasing(func, args, kwargs, new)
+
+
+def _linear_fp8_semi_sparse(input_tensor, weight_tensor, bias):
+    from torchao.ops import rowwise_scaled_linear_sparse_cutlass_f8f8
+    from torchao.quantization.quantize_.workflows.float8.float8_tensor import (
+        Float8Tensor,
+    )
+
+    if isinstance(input_tensor, Float8Tensor):
```
Comment (jcaip) [inline torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py:105]: nit: what does this conditional do?

File: torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+
+@implements([aten.mm.default, aten.addmm.default])
+def _(func, types, args, kwargs):
+    if func == aten.addmm.default:
+        bias, input_tensor, weight_tensor = args
+    else:  # aten.mm.default
+        input_tensor, weight_tensor = args
+        bias = None
+
```
Comment (jcaip) [inline torchao/quantization/quantize_/workflows/float8/float8_semi_sparse_tensor.py:135]: I think you may need to do some transpose trickery here to support mm and addmm, 

my understanding is that `linear(x, w, bias)` will return `xW^t + bias` so for mm / addmm you need to pass in a transposed weight

**Review by jcaip — COMMENTED** [review]
cc @bbeckca I think the implementation and test look good, left a couple of comments.

**Comment by bbeckca** [comment]
Abandoning in favor of #3258.
